### DISCUSSION TITLE
proxychains broken by firejail?

### DISCUSSION THREAD

**Discussion by tredondo** [netblue30/firejail#4156] [Q&A] [answered]
Not sure if this is a bug, or if I'm doing something wrong. I'm trying to firejail a Node script that needs to exit through a given IP, using proxychains. This works fine and outputs the intended IP:

```sh
$ proxychains /usr/bin/node -e "require('child_process').exec('curl https://api.ipify.org', (e, s) => console.log(s))"
[proxychains] config file found: /etc/proxychains4.conf
[proxychains] preloading /usr/lib/x86_64-linux-gnu/libproxychains.so.4
[proxychains] DLL init: proxychains-ng 4.14
[int.end.ed.ip]
```

But if I run the same command with `node` firejailed with `--noprofile`, then proxychains breaks at some point (no `[proxychains] DLL init: proxychains-ng 4.14` output), and Node ends up using my real IP:

```sh
$ proxychains firejail --noprofile /usr/bin/node -e "require('child_process').exec('curl https://api.ipify.org', (e, s) => console.log(s))"
[proxychains] config file found: /etc/proxychains4.conf
[proxychains] preloading /usr/lib/x86_64-linux-gnu/libproxychains.so.4
Parent pid 51184, child pid 51185
Child process initialized in 26.12 ms
[my.re.al.ip]

Parent is shutting down, bye...
```

[This issue](https://github.com/netblue30/firejail/issues/1093) says the combination should just work.

**Comment by tredondo** [comment]
Placing `firejail` before `proxychains` works, but I don't understand why:

```sh
$ firejail --noprofile proxychains /usr/bin/node -e "require('child_process').exec('curl https://api.ipify.org', (e, s) => console.log(s))"
Parent pid 53107, child pid 53108
Child process initialized in 6.46 ms
[proxychains] config file found: /etc/proxychains4.conf
[proxychains] preloading /usr/lib/x86_64-linux-gnu/libproxychains.so.4
[proxychains] DLL init: proxychains-ng 4.14
[int.end.ed.ip]

Parent is shutting down, bye...
```

**Comment by rusty-snake** [comment, marked as answer]
> ProxyChains is a UNIX program, that hooks network-related libc functions in DYNAMICALLY LINKED programs via a preloaded DLL (dlsym(), LD_PRELOAD)

LD_PRELOAD is ignored (and unset IIRC) by the dynamic linker for SUIDs.

https://stackoverflow.com/questions/9232892/ld-preload-with-setuid-binary

**Reply by tredondo** [reply]
![image](https://user-images.githubusercontent.com/10043369/113493752-007e4f00-947e-11eb-927d-f13eab03cfcd.png)

I'll just take it that I need to use `firejail proxychains`, in that order?

**Reply by rusty-snake** [reply]
1. Firejail is a SUID  ... (https://github.com/netblue30/firejail/blob/master/README.md)
2. ProxyChains … hooks … libc functions … via … LD_PRELOAD ... (https://github.com/rofl0r/proxychains-ng).
3. LD_PRELOAD cannot be used with SUID. (https://stackoverflow.com/questions/9232892/ld-preload-with-setuid-binary)

**Reply by tredondo** [reply]
I've read [a few](https://github.com/netblue30/firejail/discussions/4156#discussioncomment-571774) articles [about](https://www.linuxnix.com/suid-set-suid-linuxunix/) SUID and still don't understand what "Firejail is a SUID" means in this context. "In simple words users will get file owner’s permissions as well as owner UID and GID when executing a file/program/command" - OK, I'm both the user and the owner of the file (what file?), so what difference does SUID make? I'm clearly missing something.

**Reply by rusty-snake** [reply]
The SUID bit on the firejail executable (`/usr/bin/firejail`) makes firejail starting with an EUID of `0` (aka. root) instead with the EUID of the user (i.e. you) who called `firejail`.

**Comment by maltfield** [comment]
I was able to get this working with the following command

```
firejail --profile=firefox --whitelist="/home/user/.mozilla/firefox/google" proxychains firefox -no-remote -new-instance -profile "/home/user/.mozilla/firefox/google" "www.google.com"
```

I still have issues getting DNS to go through the proxychains, but otherwise it's working. I [wrote a guide](https://tech.michaelaltfield.net/2026/02/03/single-site-browser-firejail-proxychains/) with more info here:

 * https://tech.michaelaltfield.net/2026/02/03/single-site-browser-firejail-proxychains/
